File: app/src/main/java/com/github/marlonlom/demos/barcodescans/BarcodeDetectorFragment.java
package com.github.marlonlom.demos.barcodescans;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.View;

import com.google.android.gms.vision.barcode.Barcode;
import com.google.android.gms.vision.barcode.BarcodeDetector;

/**
 * Barcode detector fragment definition class.
 *
 * @author marlonlom
 */
public abstract class BarcodeDetectorFragment extends Fragment {

    private BarcodeDetector mDetector;

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        setupBarcodeDetector(getContext());
    }

    private void setupBarcodeDetector(Context appContext) {
        mDetector = new BarcodeDetector.Builder(appContext)
                .build();
    }

    protected abstract void setupView();

    protected BarcodeDetector getDetector() {
        return mDetector;
    }

    protected boolean isDetectorOperational() {
        return mDetector != null && mDetector.isOperational();
    }
}

File: app/src/main/kotlin/com/github/marlonlom/demos/barcodescans/BarcodeDetectorFragment.kt
package com.github.marlonlom.demos.barcodescans

import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.View

import com.google.android.gms.vision.barcode.Barcode
import com.google.android.gms.vision.barcode.BarcodeDetector

/**
 * Barcode detector fragment definition class.

 * @author marlonlom
 */
abstract class BarcodeDetectorFragment : Fragment() {

    protected var detector: BarcodeDetector? = null
        private set

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setupBarcodeDetector(context)
    }

    private fun setupBarcodeDetector(appContext: Context) {
        detector = BarcodeDetector.Builder(appContext)
                .build()
    }

    protected abstract fun setupView()

    protected val isDetectorOperational: Boolean
        get() = detector != null && detector!!.isOperational
}
